bazel-build.el

```diff
@@ -0,0 +1,200 @@
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+;;; Commentary:
+;;
+;; This package provides commands to build and run code using Bazel.
+;; It defines interactive commands `bazel-build' and `bazel-run' which
+;; perform completion of available Bazel targets.
+;;
+;;; Code:
+
+(require 'cl-lib)
+
+(defun bazel-build ()
+  "Find and build a Bazel target."
+  (interactive)
+  (bazel-build--run-bazel-command "build"))
+
+(defun bazel-run ()
+  "Find and run a Bazel target."
+  (interactive)
+  (bazel-build--run-bazel-command "run"))
+
+(defun bazel-test ()
+  "Find and run a Bazel test target."
+  (interactive)
+  (bazel-build--run-bazel-command "test"))
+
+(defun bazel-build--run-bazel-command (command)
+  "Run Bazel tool with given COMMAND, e.g. build or run."
+  (let* ((file-name (buffer-file-name))
+         (workspace-root (bazel-build--find-workspace-root file-name))
+         (initial-input (bazel-build--suggest-initial-input file-name workspace-root))
+         (target (completing-read
+				  (format "bazel %s " command) ; prompt
+                  #'bazel-build--completions   ; collection
+                  nil                          ; predicate
+                  nil                          ; require-match
+                  initial-input)))             ; initial-input
+	(compile
+	 (mapconcat #'shell-quote-argument (list "bazel" command target) " "))))
+
+(defun bazel-build--find-workspace-root (file-name)
+  "Find the root of the Bazel workspace containing FILE-NAME."
+  (let ((workspace-root (locate-dominating-file file-name "WORKSPACE")))
+    (if workspace-root (expand-file-name workspace-root)
+      (error "Not in a Bazel workspace"))))
+
+(defun bazel-build--suggest-initial-input (file-name workspace-root)
+  "Suggest initial input to the interactive completion.
+FILE-NAME is the file-name of the current buffer.  WORKSPACE-ROOT is
+the root of the Bazel workspace."
+  (concat "//" (directory-file-name (bazel-build--extract-package-name file-name workspace-root))))
+
+(defconst bazel-build--recognized-build-file-names
+  '("BUILD.bazel" "BUILD")
+  "Names files Bazel will look in to find build rules.")
+
+(defun bazel-build--extract-package-name (file-name workspace-root)
+  "Extract the nearest Bazel package for FILE-NAME under WORKSPACE-ROOT."
+  (file-relative-name
+   (expand-file-name
+	(or
+	 (cl-some
+	  (lambda (build-name) (locate-dominating-file file-name build-name)) bazel-build--recognized-build-file-names)
+	 workspace-root))
+   workspace-root))
+
+(defun bazel-build--completions (string pred mode)
+  "Programmed completion for arguments to 'bazel-build', 'bazel-run', etc.
+STRING is the string to be completed by this rule.  PRED is a
+predicate which may be used to filter possible completions.  MODE
+specifies the completion mode."
+  (let* ((string-split (bazel-build--extract-special-prefix string))
```

Comment:
It might be helpful to inline a few of the helper functions and restructure the code to make it easier to read. IIUC, what you want is something like "if the user entered something starting with //, then complete package names, else complete targets starting from the current package". I guess this is what the code implements, but it's hard to tell.